### PR TITLE
Watchlist settings

### PR DESCRIPTION
**Pull request by IvanIvanoff**
#### Summary
1. Restructure the Signal.Validation into two modules, so one of them is more general and is used in the watchlists.
2. Add watchlist settings per user. The behavior is the following:
- If the user or the watchlist creator does not have settings - return default settings
- If the user does not have settings and the creator does have - return the creator's settings
- If the user has settings - return his or her settings

Examples:
```graphql
mutation {
        updateWatchlistSettings(
          id: 10,
          settings: {
            pageSize: 50
            timeWindow: "180d"
            tableColumns: <json map with arbitrary keys>
          }) {
            pageSize
            tableColumns
            timeWindow
          }
```

[//]: # (#### Related PRs)

[//]: # (#### Additional deploy notes)

[//]: # (#### Screenshots)

### CODE REVIEW

File: lib/sanbase_web/graphql/resolvers/user/user_list_resolver.ex [hunk 1]
```diff
@@ -129,6 +139,23 @@ defmodule SanbaseWeb.Graphql.Resolvers.UserListResolver do
...
+  def update_watchlist_settings(_root, %{id: watchlist_id, settings: settings}, %{
+        context: %{auth: %{current_user: current_user}}
+      }) do
+    UserList.Settings.update_or_create_settings(watchlist_id, current_user.id, settings)
+    |> case do
+      {:ok, %{settings: settings}} ->
+        {:ok, settings}
+
+      {:error, changeset} ->
+        # {:error,
```
Comment (kalkov) [inline lib/sanbase_web/graphql/resolvers/user/user_list_resolver.ex]: These are leftovers probably?

File: priv/repo/migrations/20190710140802_add_watchlist_settings_table.exs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+defmodule Sanbase.Repo.Migrations.AddWatchlistSettingsTable do
```
Comment (kalkov) [inline priv/repo/migrations/20190710140802_add_watchlist_settings_table.exs]: Two migrations with same content but diff names?

**Comment by kalkov** [comment]
Can we also have tests for validations?

File: priv/repo/migrations/20190710140802_add_watchlist_settings_table.exs (hunk 2 above)
Comment (IvanIvanoff) [reply priv/repo/migrations/20190710140802_add_watchlist_settings_table.exs]: Hmm, I couldn't find the migration and created a new one. Obviously it was where it should be :smiley:

File: lib/sanbase_web/graphql/resolvers/user/user_list_resolver.ex (hunk 1 above)
Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/resolvers/user/user_list_resolver.ex]: Yeah, I need to do some cleaning

File: lib/sanbase/user_lists/settings.ex [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+  The settings are stored per user. One watchlist can have many settings for many
+  users. When a user requires the settings for a watchlist the resolution process
+  is the first one that matches:
+  1. If the requesting user has defined settings - return them.
+  2. If the watchlist craetor has defined settings - return them.
+  3. Return the default settings
+  """
+
+  defmodule WatchlistSettings do
```
Comment (tspenov) [inline lib/sanbase/user_lists/settings.ex]: why nested modules ?

Comment (IvanIvanoff) [reply lib/sanbase/user_lists/settings.ex]: no particular reason. I'm just trying how it looks. And this module won't be used by anyone else

File: lib/sanbase_web/graphql/resolvers/user/user_list_resolver.ex [hunk 4]
```diff
@@ -129,6 +139,23 @@ defmodule SanbaseWeb.Graphql.Resolvers.UserListResolver do
...
+    |> case do
+      {:ok, %{settings: settings}} ->
+        {:ok, settings}
+
+      {:error, %Ecto.Changeset{} = changeset} ->
+        {:error,
+         message: "Cannot update watchlist settings",
+         details: Sanbase.Utils.ErrorHandling.changeset_errors_to_str(changeset)}
+
+        # {:error, inspect(changeset)}
```
Comment (tspenov) [inline lib/sanbase_web/graphql/resolvers/user/user_list_resolver.ex]: leftover

File: lib/sanbase/user_lists/settings.ex [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+  use Ecto.Schema
+
+  import Ecto.Query
+  import Ecto.Changeset
+
+  alias Sanbase.Repo
+  alias Sanbase.Auth.User
+  alias Sanbase.UserList
+  alias __MODULE__.WatchlistSettings
+
```
Comment (tspenov) [inline lib/sanbase/user_lists/settings.ex]: shouldn't we set `@primary_key false` as per: https://hexdocs.pm/ecto/Ecto.Schema.html#module-primary-keys ?

File: priv/repo/migrations/20190710140802_add_watchlist_settings_table.exs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+  use Ecto.Migration
+
+  def change do
+    create table(:watchlist_settings) do
```
Comment (tspenov) [inline priv/repo/migrations/20190710140802_add_watchlist_settings_table.exs]: I think we need `primary_key: false` here and `primary_key: true` on each of the fields making the composite primary_key

File: lib/sanbase/user_lists/settings.ex (hunk 5 above)
Comment (IvanIvanoff) [reply lib/sanbase/user_lists/settings.ex]: The table still has primary key `id`. I probably should remote the `primary_key: true`

File: lib/sanbase_web/graphql/schema/queries/watchlist_queries.ex [hunk 7]
```diff
@@ -122,6 +122,13 @@ defmodule SanbaseWeb.Graphql.Schema.WatchlistQueries do
       resolve(&UserListResolver.update_user_list/3)
     end
 
+    field :update_watchlist_settings, :watchlist_settings do
+      arg(:id, non_null(:integer))
+      arg(:settings, :watchlist_settings_input_object)
```
Comment (tspenov) [inline lib/sanbase_web/graphql/schema/queries/watchlist_queries.ex]: non_null ?

File: lib/sanbase/user_lists/settings.ex [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    settings =
+      get_settings(ul.id, user.id) ||
+        get_settings(ul.id, ul.user_id) ||
+        WatchlistSettings.default_settings()
+
+    {:ok, settings}
+  end
+
+  def settings_for(%UserList{} = ul, _) do
+    get_settings(ul.id, ul.user_id)
```
Comment (tspenov) [inline lib/sanbase/user_lists/settings.ex]: That id getting access to settings of watchlist from not-logged in user: `UserList.Settings.settings_for(watchlist, nil)`. Should we check at least that watchlist id public ?

File: lib/sanbase/user_lists/settings.ex (hunk 5 above)
Comment (IvanIvanoff) [reply lib/sanbase/user_lists/settings.ex]: Hmm. Or better to rework the migration and indeed have this pair as primary key

File: lib/sanbase/user_lists/settings.ex [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+  exists:
+  1. User's settings for that watchlist
+  2. The watchlist's creator's settings for that watchlist
+  3. Default settings
+  """
+  @spec settings_for(%UserList{}, %User{} | nil) :: {:ok, WatchlistSettings.t()}
+  def settings_for(%UserList{} = ul, %User{} = user) do
+    settings =
+      get_settings(ul.id, user.id) ||
+        get_settings(ul.id, ul.user_id) ||
```
Comment (tspenov) [inline lib/sanbase/user_lists/settings.ex]: This is another user's watchlist. Shouldn't this check if watchlist is public ?

File: lib/sanbase/user_lists/settings.ex (hunk 3 above)
Comment (tspenov) [reply lib/sanbase/user_lists/settings.ex]: yes, but looks confusing at first sight. Maybe just place it in the same file but not nested like we did for triggers :?

File: lib/sanbase_web/graphql/schema/queries/watchlist_queries.ex (hunk 7 above)
Comment (IvanIvanoff) [reply lib/sanbase_web/graphql/schema/queries/watchlist_queries.ex]: :+1:

File: lib/sanbase/user_lists/settings.ex (hunk 8 above)
Comment (IvanIvanoff) [reply lib/sanbase/user_lists/settings.ex]: `settings` a field in the user_list type and is accessed only if the user is granted access to the watchlist.

File: lib/sanbase/user_lists/settings.ex (hunk 9 above)
Comment (IvanIvanoff) [reply lib/sanbase/user_lists/settings.ex]: You cannot reach here as you'll be stopped by graphql - settings is being resolved only if you're given access to it

File: lib/sanbase/user_lists/settings.ex (hunk 3 above)
Comment (IvanIvanoff) [reply lib/sanbase/user_lists/settings.ex]: If not nested the module name will be less readable (UserList.Settings.WatchlistSettings). I can rename it to `Settings.EmbededSchema` for example and separate it?

Comment (tspenov) [reply lib/sanbase/user_lists/settings.ex]: I can live with how it is now.

**Comment by tspenov** [comment]
@IvanIvanoff build is failing.

**Pull request merged**